### PR TITLE
refactored some stuff

### PR DESCRIPTION
**Pull request by sandrodevdariani**
1.  **https://board.desearch.ai/Desearch-Board-14db8bfdbde98089be4ec47fc9be346e?p=152b8bfdbde980c8833eec02a9860a1a&pm=s**

Implement  Exponential Penalties in **streaming_penatly.py** file and update **scraper_validator.py** to using that new  Exponential Penalties function.

2. Update **compute_rewards_and_penalties** function  in **neurons/validators/scraper_validator.py** to call **update_moving_averaged_score** if query is synthetic, otherwise **scattered_rewards = rewards**

3. Add details log in Organic scoring

4. Removed unnecessary imports 

5. Add Exponential Retry Mechanism OpenAI API Calls for **dataset.py**

6. add include_summary param for **get_system_summary_relevance_scoring_template** function

### CODE REVIEW

File: neurons/validators/organic_query_state.py [hunk 1]
```diff
@@ -4,25 +4,30 @@
 from datura.dataset.date_filters import DateFilter, DateFilterType
 from datetime import datetime
 import pytz
-import bittensor as bt
+import logging
```
Comment (Chkhikvadze) [inline neurons/validators/organic_query_state.py:7]: Why did you remove bittensor as bt?

File: neurons/validators/penalty/streaming_penalty.py [hunk 2]
```diff
@@ -29,23 +46,41 @@ def calculate_penalties(
...
             for token_count in token_counts:
                 if token_count > MAX_TOKENS_PER_CHUNK:
-                    penalty = (
-                        token_count - MAX_TOKENS_PER_CHUNK
-                    ) * PENALTY_PER_EXCEEDING_TOKEN
+                    penalty = (token_count - MAX_TOKENS_PER_CHUNK) * PENALTY_PER_EXCEEDING_TOKEN
+                    token_penalty = min(1.0, token_penalty + penalty)
+
+            #Time-Based Penalty Calculation
+            if response.elapsed_time is not None:
```
Comment (okradze) [inline neurons/validators/penalty/streaming_penalty.py:66]: use response.dendrite.process_time instead of adding new field elapsed_time

File: neurons/validators/utils/prompts.py [hunk 3]
```diff
@@ -261,27 +261,31 @@ def clean_template(template):
...
+def get_system_summary_relevance_scoring_template(tools: List[str], include_summary: bool = True):
+    """Generate the system message for the Summary Relevance Scoring prompt based on tools and summary inclusion."""
 
     links_header_name = ""
     summary_header_name = ""
 
     if "Twitter Search" in tools:
         links_header_name = "**Key Tweets**"
-        summary_header_name = "**Twitter Summary**"
+        summary_header_name = "**Twitter Summary**" if include_summary else ""
```
Comment (Chkhikvadze) [inline neurons/validators/utils/prompts.py:272]: why did you add it? it should be always here, is not correct

File: requirements.txt [hunk 4]
```diff
@@ -35,3 +35,4 @@ openai
 python-dateutil
 pytz
 faker==25.9.1
+asyncio
```
Comment (Chkhikvadze) [inline requirements.txt:38]: Why did you add it? do we need it?

File: neurons/validators/utils/prompts.py [hunk 5]
```diff
@@ -261,27 +261,31 @@ def clean_template(template):
     return "\n".join(cleaned_lines)
 
 
-def get_system_summary_relevance_scoring_template(tools: List[str]):
-    """Generate the system message for the Summary Relevance Scoring prompt based on tools"""
+def get_system_summary_relevance_scoring_template(tools: List[str], include_summary: bool = True):
```
Comment (okradze) [inline neurons/validators/utils/prompts.py:264]: keep this as it was
it's called only in one place and no need for flag

File: neurons/validators/penalty/streaming_penalty.py [hunk 6]
```diff
@@ -5,11 +5,28 @@
 import bittensor as bt
 from datura.protocol import ScraperStreamingSynapse
 import tiktoken
+import math
 
 
 MAX_TOKENS_PER_CHUNK = 2
 PENALTY_PER_EXCEEDING_TOKEN = 0.01
 
+HARD_TIMEOUT = 10.0  # Hard timeout in seconds
```
Comment (Chkhikvadze) [inline neurons/validators/penalty/streaming_penalty.py:14]: is wrong, you need to control it by max_execution_time which is used when dendrite is called.

Check models:
Nova
Orbit
Horizon

File: neurons/validators/penalty/streaming_penalty.py [hunk 7]
```diff
@@ -29,23 +46,41 @@ def calculate_penalties(
...
-                        token_count - MAX_TOKENS_PER_CHUNK
-                    ) * PENALTY_PER_EXCEEDING_TOKEN
+                    penalty = (token_count - MAX_TOKENS_PER_CHUNK) * PENALTY_PER_EXCEEDING_TOKEN
+                    token_penalty = min(1.0, token_penalty + penalty)
+
+            #Time-Based Penalty Calculation
+            if response.elapsed_time is not None:
+                time_penalty = calculate_time_penalty(
+                    elapsed_time=response.elapsed_time,
+                    limit=HARD_TIMEOUT,
```
Comment (Chkhikvadze) [inline neurons/validators/penalty/streaming_penalty.py:69]: is wrong to define Hard_timeout, need to be used from response max_execution_time (Models)

File: neurons/validators/utils/prompts.py [hunk 8]
```diff
@@ -261,27 +261,31 @@ def clean_template(template):
...
-        summary_header_name = "**Twitter Summary**"
+        summary_header_name = "**Twitter Summary**" if include_summary else ""
     elif "Hacker News Search" in tools:
         links_header_name = "**Key News**"
-        summary_header_name = "**Hacker News Summary**"
+        summary_header_name = "**Hacker News Summary**" if include_summary else ""
     elif "Reddit Search" in tools:
         links_header_name = "**Key Posts**"
-        summary_header_name = "**Reddit Summary**"
+        summary_header_name = "**Reddit Summary**" if include_summary else ""
```
Comment (Chkhikvadze) [inline neurons/validators/utils/prompts.py:278]: no need, why?

**Review by Chkhikvadze — CHANGES_REQUESTED** [review]
